deeppavlov/pipeline_manager/pipegen.py

```diff
@@ -0,0 +1,273 @@
+            save_path: path to folder with pipelines checkpoints.
+            mode: working mode of generator, can be 'random' or 'grid'.
+            sample_num: determines the number of generated pipelines, if hyper_search == random.
+            test_mode: trigger that determine logic of changing save and loads paths in config.
+    """
+
+    def __init__(self,
+                 config: Union[str, Dict, Path],
+                 save_path: Union[str, Path],
+                 mode: str = 'random',
+                 sample_num: int = 10,
+                 test_mode: bool = False) -> None:
+        """ Initialize generator with input params. """
+        if isinstance(config, dict):
+            self.main_config = deepcopy(config)
+        else:
+            self.main_config = deepcopy(read_json(config))
+
+        self.dataset_reader = self.main_config.pop("dataset_reader")
+        if not isinstance(self.main_config["dataset_iterator"], dict):
+            raise ConfigError("Dataset iterator must be one for hole experiment.")
+
+        self.train_config = self.main_config.pop("train")
+        self.chainer = self.main_config.pop('chainer')
+        self.structure = self.chainer['pipe']
+
+        if mode in ['random', 'grid']:
+            self.mode = mode
+        else:
+            raise ConfigError(f"'{mode} search' not implemented. Only 'grid' and 'random' search are available.")
+
+        self.test_mode = test_mode
+        self.save_path = save_path
+        self.length = None
+        self.pipes = []
+        self.N = sample_num
+
+        self._check_component_name()
+        self.get_len()
+        self.enumerator = self.pipeline_enumeration()
+        self.generator = self.pipeline_gen()
+
+    def _check_component_name(self) -> None:
+        """ Checks incoming config for the presence of a "component_name" key in the component description dict. """
+        for i, component in enumerate(self.structure):
+            for j, example in enumerate(component):
+                if example is not None:
+                    if "component_name" not in example.keys():
+                        raise ConfigError("The pipeline element in config file, on position {0} and with number {1}"
+                                          "don't contain the 'component_name' key.".format(i + 1, j + 1))
+
+    def get_len(self) -> None:
+        """ Calculate the length of generator. """
+        self.enumerator = self.pipeline_enumeration()
+        generator = self.pipeline_gen()
+        self.length = len(list(generator))
+        self.pipes = []
+        del generator
+
+    def __len__(self):
+        return self.length
+
+    def pipeline_enumeration(self) -> Iterator:
+        """
+        Creates a primary set of pipelines using a self.main_config attribute.
+
+        Returns:
+            iterator of primary set of pipelines
+        """
+        for components in self.structure:
+            self.pipes.append(components)
```

Comment:
every time this method is called, `self.pipes` grows. And it's called at least twice: in `__init__()` and in `get_len()`